File: src/+sem/+util/Units.m
classdef Units
    %Units Mode-aware conversions between cell units and DAQ volts.
    %   Lifted from the lab's whole-cell patch software (patchclamp.analysis.Units)
    %   and restyled for this repo. There are two physical signal paths, each
    %   with its own scaling:
    %
    %     1. Command path (DAQ AO -> 700B EXT COMMAND)
    %          VC: gain.commandVcMvPerV  mV per V at the DAQ AO  (typ. 20)
    %          IC: gain.commandIcPaPerV  pA per V at the DAQ AO  (typ. 400)
    %        These are PHYSICAL CONSTANTS of the 700B EXT COMMAND input.
    %
    %     2. Scaled-output path (700B SCALED OUTPUT -> DAQ AI)
    %          VC: gain.scaledVcPaPerV   pA per V at the AI       (typ. 1000)
    %          IC: gain.scaledIcMvPerMv  mV at the AI per mV at the cell (typ. 20)
    %        These are set on the front panel / telegraph; snapshot them into
    %        every trial's metadata.ephys.gains.
    %
    %   All conversions live in this file so they can be audited in one place.
    %   Anywhere else in the MATLAB code that does its own conversion is a bug.
    %   (The Python analysis package mirrors these in units.py — change both
    %   together.)

    methods (Static)
        function gain = gainFromConfig(ephysCfg)
            %gainFromConfig Build the gain struct from a config ephys section.
            gain = struct( ...
                'commandVcMvPerV', sem.util.configField(ephysCfg, 'commandVcMvPerV', 20), ...
                'commandIcPaPerV', sem.util.configField(ephysCfg, 'commandIcPaPerV', 400), ...
                'scaledVcPaPerV',  sem.util.configField(ephysCfg, 'scaledVcPaPerV', 1000), ...
                'scaledIcMvPerMv', sem.util.configField(ephysCfg, 'scaledIcMvPerMv', 20));
        end

        function volts = commandCellToDaqVolts(cellValue, mode, gain)
            %commandCellToDaqVolts Cell-unit command (mV in VC, pA in IC) -> DAQ AO volts.
            sem.util.Units.checkMode(mode);
            if strcmp(char(mode), 'VC')
                volts = cellValue ./ gain.commandVcMvPerV;
            else
                volts = cellValue ./ gain.commandIcPaPerV;
            end
        end

        function cellValue = commandDaqVoltsToCell(volts, mode, gain)
            %commandDaqVoltsToCell Inverse of commandCellToDaqVolts.
            sem.util.Units.checkMode(mode);
            if strcmp(char(mode), 'VC')
                cellValue = volts .* gain.commandVcMvPerV;
            else
                cellValue = volts .* gain.commandIcPaPerV;
            end
        end

        function cellValue = scaledDaqVoltsToCell(volts, mode, gain)
            %scaledDaqVoltsToCell AI sample (DAQ volts) -> cell units (pA in VC, mV in IC).
            %   IC ratio is dimensionless (mV out per mV at the cell); AI is in
            %   volts, so *1000 converts V to mV before dividing.
            sem.util.Units.checkMode(mode);
            if strcmp(char(mode), 'VC')
                cellValue = volts .* gain.scaledVcPaPerV;
            else
                cellValue = volts .* (1000 ./ gain.scaledIcMvPerMv);
            end
        end

        function volts = scaledCellToDaqVolts(cellValue, mode, gain)
            %scaledCellToDaqVolts Inverse of scaledDaqVoltsToCell.
            %   Used by MockEphysDAQ to synthesise an AI trace in volts from a
            %   known cell-unit ground truth.
            sem.util.Units.checkMode(mode);
            if strcmp(char(mode), 'VC')
                volts = cellValue ./ gain.scaledVcPaPerV;
            else
                volts = cellValue .* (gain.scaledIcMvPerMv ./ 1000);
            end
        end
    end

    methods (Static, Access = private)
        function checkMode(mode)
            if ~ismember(char(mode), {'VC', 'IC'})
                error('sem:util:Units:badMode', ...
                    'mode must be ''VC'' or ''IC''; got ''%s''.', char(mode));
            end
        end
    end
end

File: analysis/src/slice_ephys_analysis/units.py
"""Cell-unit <-> DAQ-volt conversions. Python mirror of sem.util.Units.

This module and MATLAB's sem/util/Units.m implement the same contract —
change them together. Raw aiData is saved in DAQ VOLTS; every conversion to
pA/mV happens here, using the per-trial gains snapshot from
metadata.ephys.gains (single source of truth).

Gain struct fields:
  commandVcMvPerV   DAQ AO -> 700B EXT COMMAND, VC (physical constant, 20)
  commandIcPaPerV   DAQ AO -> 700B EXT COMMAND, IC (physical constant, 400)
  scaledVcPaPerV    700B SCALED OUTPUT -> DAQ AI, VC (front panel, typ. 1000)
  scaledIcMvPerMv   700B SCALED OUTPUT -> DAQ AI, IC (front panel, typ. 20)
"""

from __future__ import annotations

import numpy as np


def _check_mode(mode: str) -> str:
    if mode not in ("VC", "IC"):
        raise ValueError(f"mode must be 'VC' or 'IC'; got {mode!r}")
    return mode


def scaled_daq_volts_to_cell(volts, mode: str, gain: dict):
    """AI sample (DAQ volts) -> cell units (pA in VC, mV in IC)."""
    _check_mode(mode)
    v = np.asarray(volts, dtype=float)
    if mode == "VC":
        return v * float(gain["scaledVcPaPerV"])
    return v * (1000.0 / float(gain["scaledIcMvPerMv"]))


def scaled_cell_to_daq_volts(cell_value, mode: str, gain: dict):
    """Inverse of scaled_daq_volts_to_cell."""
    _check_mode(mode)
    c = np.asarray(cell_value, dtype=float)
    if mode == "VC":
        return c / float(gain["scaledVcPaPerV"])
    return c * (float(gain["scaledIcMvPerMv"]) / 1000.0)


def command_cell_to_daq_volts(cell_value, mode: str, gain: dict):
    """Command in cell units (mV in VC, pA in IC) -> DAQ AO volts."""
    _check_mode(mode)
    c = np.asarray(cell_value, dtype=float)
    if mode == "VC":
        return c / float(gain["commandVcMvPerV"])
    return c / float(gain["commandIcPaPerV"])


def command_daq_volts_to_cell(volts, mode: str, gain: dict):
    """Inverse of command_cell_to_daq_volts."""
    _check_mode(mode)
    v = np.asarray(volts, dtype=float)
    if mode == "VC":
        return v * float(gain["commandVcMvPerV"])
    return v * float(gain["commandIcPaPerV"])
